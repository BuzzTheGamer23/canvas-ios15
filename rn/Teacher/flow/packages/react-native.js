// @flow
export type PanResponderGestureState = {
  stateID: number,
  moveX: number,
  moveY: number,
  x0: number,
  y0: number,
  dx: number,
  dy: number,
  vx: number,
  vy: number,
}
