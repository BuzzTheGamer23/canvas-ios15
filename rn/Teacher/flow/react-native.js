//
// Copyright (C) 2017-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow
/* eslint-disable */

// define basic style prop types
// the internal ones are the best we have
import {
  type ____ViewStyleProp_Internal,
  type ____TextStyleProp_Internal,
  type ____ImageStyleProp_Internal,
} from 'react-native/Libraries/StyleSheet/StyleSheetTypes'

declare type ViewStyleProp = ____ViewStyleProp_Internal
declare type TextStyleProp = ____TextStyleProp_Internal
declare type ImageStyleProp = ____ImageStyleProp_Internal
