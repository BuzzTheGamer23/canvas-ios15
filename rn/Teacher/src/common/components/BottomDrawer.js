// @flow

import React, { Component } from 'react'
import {
  StyleSheet,
  Dimensions,
  Animated,
  View,
  Keyboard,
} from 'react-native'
import { connect } from 'react-redux'
import Interactable from 'react-native-interactable'

let { height, width } = Dimensions.get('window')

const CLOSED_PANEL_HEIGHT = 70
const BOTTOM_FLUFF = 1024

type Props = SnapState & {
  containerWidth?: number,
  containerHeight?: number,
  children?: React.Element<*>,
  setDrawerSnap: ((snap: Snap) => void),
}

type State = {
  height: number,
  width: number,
  bottomPadding: number,
}

export class BottomDrawer extends Component<any, Props, State> {
  props: Props
  state: State
  _deltaY: Animated.Value
  drawer: InteractableView

  constructor (props: Props) {
    super(props)

    this.state = {
      height: props.containerHeight || height,
      width: props.containerWidth || width,
      currentSnap: 2,
      bottomPadding: height * 0.8 - CLOSED_PANEL_HEIGHT,
    }

    this._deltaY = new Animated.Value(this.state.height - CLOSED_PANEL_HEIGHT)
  }

  componentWillReceiveProps (nextProps: Props) {
    if (nextProps.containerHeight !== this.props.containerHeight || nextProps.containerWidth !== this.props.containerWidth) {
      this.setState({ height: nextProps.containerHeight, width: nextProps.containerWidth }, () => {
        this.drawer.snapTo({ index: this.props.currentSnap })
      })
    }
  }

  componentDidUpdate (prevProps: Props) {
    if (prevProps.currentSnap !== this.props.currentSnap) {
      this.drawer.snapTo({ index: this.props.currentSnap })
    }
  }

  open () {
    if (this.state.currentSnap === 2) {
      this.drawer.snapTo({ index: 1 })
    }
  }

  onSnap = (e: any) => {
    if (e.nativeEvent.index !== this.props.currentSnap) {
      this.props.setDrawerSnap(e.nativeEvent.index)
    }
  }

  getSnapPoints = () => {
    const fullScreen = { y: 44 }
    // open and closed must be larger than fullscreen
    // this only occurs when the view containing the
    // drawer is smaller than 44 + CLOSED_PANEL_HEIGHT
    // which happens in the unit tests
    const open = { y: Math.max(fullScreen.y + 1, this.state.height * 0.6 - CLOSED_PANEL_HEIGHT) }
    const closed = { y: Math.max(fullScreen.y + 2, this.state.height - CLOSED_PANEL_HEIGHT) }

    if (Keyboard.isVisible) {
      return [fullScreen, closed]
    } else {
      return [fullScreen, open, closed]
    }
  }

  render () {
    let width = Math.min(700, this.state.width)
    let position = (this.state.width - width) / 2

    const snap = this.getSnapPoints().map(point => point.y)
    const clamped = [...snap.slice(0, snap.length - 1), snap[1]].map(y => y + BOTTOM_FLUFF)
    const clamp = this._deltaY.interpolate({
      inputRange: snap,
      outputRange: clamped,
    })
    return (
      <Interactable.View
        ref={(e) => { this.drawer = e }}
        onSnap={this.onSnap}
        verticalOnly={true}
        snapPoints={this.getSnapPoints()}
        initialPosition={this.getSnapPoints()[this.props.currentSnap]}
        animatedValueY={this._deltaY}
        style={[styles.panelContainer, { left: position, right: position }]}
        onLayout={this.onLayout}
      >
        <View style={styles.drawerBackground} />
        <Animated.View
          style={[styles.panel, {
            height: this.state.height,
            paddingBottom: clamp,
          }]}
        >
          <View style={styles.handleWrapper}>
            <View style={styles.handle} />
          </View>
          {this.props.children}
        </Animated.View>
      </Interactable.View>
    )
  }
}

const DRAWER_SNAP_ACTION = 'com.teacher.drawer.set-snap'
const DRAWER_OPEN_ACTION = 'com.teacher.drawer.open'
type DrawerAction
  = { type: 'com.teacher.drawer.open' }
  | { type: 'com.teacher.drawer.set-snap', snap: Snap }

export const createSnapAction = (snap: Snap): DrawerAction => ({ type: DRAWER_SNAP_ACTION, snap: snap })

export function drawer (state: SnapState = { currentSnap: 2 }, action: DrawerAction): SnapState {
  switch (action.type) {
    case DRAWER_OPEN_ACTION:
      if (state.currentSnap === 2) {
        return { ...state, currentSnap: 1 }
      }
      break
    case DRAWER_SNAP_ACTION:
      return { ...state, currentSnap: action.snap }
  }
  return state
}

export const DrawerActions: any = {
  resetDrawer: () => createSnapAction(2),
  setDrawerSnap: createSnapAction,
  openDrawer: () => ({ type: DRAWER_OPEN_ACTION }),
}

export function mapStateToProps (state: AppState): SnapState {
  return state.drawer
}

const Connected = connect(mapStateToProps, DrawerActions)(BottomDrawer)
export default (Connected: any)

const styles = StyleSheet.create({
  panelContainer: {
    position: 'absolute',
    top: 0,
    bottom: -BOTTOM_FLUFF,
    left: 0,
    right: 0,
    maxWidth: 700,
  },
  panel: {
    paddingTop: 6,
    paddingBottom: 0,
    flex: 1,
  },
  handleWrapper: {
    position: 'absolute',
    top: 18,
    left: 0,
    right: 0,
    alignItems: 'center',
    marginVertical: 3,
  },
  handle: {
    width: 40,
    height: 4,
    backgroundColor: 'darkgray',
    borderRadius: 5,
    marginTop: 4,
    marginBottom: 8,
  },
  drawerBackground: {
    shadowColor: '#000000',
    shadowOffset: { width: 0, height: 0 },
    shadowRadius: 3,
    shadowOpacity: 0.3,
    backgroundColor: '#FFFFFF',
    borderTopLeftRadius: 12,
    borderTopRightRadius: 12,
    position: 'absolute',
    margin: 0,
    padding: 0,
    top: 20,
    bottom: 0,
    left: 0,
    right: 0,
  },
})
