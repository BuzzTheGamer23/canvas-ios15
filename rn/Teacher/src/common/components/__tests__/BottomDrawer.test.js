// @flow
import React from 'react'
import { Text } from 'react-native'
import { BottomDrawer } from '../BottomDrawer'
import renderer from 'react-test-renderer'
import setProps from '../../../../test/helpers/setProps'

jest.mock('react-native-interactable', () => ({
  View: 'Interactable.View',
}))

describe('BottomDrawer', () => {
  it('renders any children', () => {
    let tree = renderer.create(
      <BottomDrawer currentSnap={2}>
        <Text>Yo yo yo</Text>
      </BottomDrawer>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('sets the new currentSnap when onSnap is called', () => {
    const setDrawerSnap = jest.fn()
    let tree = renderer.create(
      <BottomDrawer setDrawerSnap={setDrawerSnap}>
        <Text>yo yo yo</Text>
      </BottomDrawer>
    )
    let instance = tree.getInstance()
    instance.onSnap({
      nativeEvent: {
        index: 1,
      },
    })
    expect(setDrawerSnap).toHaveBeenCalledWith(1)
  })

  it('sets the new height and width state when props change', () => {
    let tree = renderer.create(
      <BottomDrawer containerHeight={100} containerWidth={100}>
        <Text>yo yo yo</Text>
      </BottomDrawer>
    )
    let instance = tree.getInstance()
    expect(instance.state.height).toEqual(100)
    expect(instance.state.width).toEqual(100)

    setProps(tree, { containerHeight: 200, containerWidth: 200 })
    instance = tree.getInstance()
    expect(instance.state.height).toEqual(200)
    expect(instance.state.width).toEqual(200)
  })

  it('calls snapTo when currentSnap is 2', () => {
    let tree = renderer.create(
      <BottomDrawer>
        <Text>Yo yo yo</Text>
      </BottomDrawer>
    )
    let instance = tree.getInstance()
    instance.drawer = { snapTo: jest.fn() }
    instance.open()
    expect(instance.drawer.snapTo).toHaveBeenCalledWith({ index: 1 })
  })

  it('doesnt call snapTo when the currentSnap is not 2', () => {
    let tree = renderer.create(
      <BottomDrawer>
        <Text>yo yo yo</Text>
      </BottomDrawer>
    )
    let instance = tree.getInstance()
    instance.setState({ currentSnap: 0 })
    instance.drawer = { snapTo: jest.fn() }
    instance.open()
    expect(instance.drawer.snapTo).not.toHaveBeenCalled()
  })

  it('returns 3 snap points from getSnapPoints', () => {
    let tree = renderer.create(
      <BottomDrawer>
        <Text>yo yo yo</Text>
      </BottomDrawer>
    )
    let instance = tree.getInstance()
    let snapPoints = instance.getSnapPoints()
    expect(snapPoints[0].y).toBeLessThan(snapPoints[1].y)
    expect(snapPoints[1].y).toBeLessThan(snapPoints[2].y)
    expect(snapPoints.length).toEqual(3)
  })
})
