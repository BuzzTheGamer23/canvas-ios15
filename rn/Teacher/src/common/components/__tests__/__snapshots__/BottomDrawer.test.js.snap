// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BottomDrawer renders any children 1`] = `
<Interactable.View
  animatedValueY={1281}
  initialPosition={
    Object {
      "y": 1281,
    }
  }
  onLayout={undefined}
  onSnap={[Function]}
  snapPoints={
    Array [
      Object {
        "y": 213.8,
      },
      Object {
        "y": 747.4,
      },
      Object {
        "y": 1281,
      },
    ]
  }
  style={
    Array [
      Object {
        "bottom": 0,
        "left": 0,
        "maxWidth": 700,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      Object {
        "left": 25,
        "right": 25,
      },
    ]
  }
  verticalOnly={true}
>
  <View
    style={
      Object {
        "backgroundColor": "#fff",
        "borderTopLeftRadius": 20,
        "borderTopRightRadius": 20,
        "flex": 1,
        "height": 1334,
        "paddingBottom": 1281,
        "paddingTop": 16,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 0.3,
        "shadowRadius": 3,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#e3e3e3",
            "borderRadius": 5,
            "height": 4,
            "marginBottom": 8,
            "marginTop": 4,
            "width": 40,
          }
        }
      />
    </View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      Yo yo yo
    </Text>
  </View>
</Interactable.View>
`;
